src/events/backup.py

```diff
@@ -139,6 +149,128 @@ def _on_list_backups_action(self, event: ActionEvent):
         event.log(output)
         event.set_results({"backups": json.dumps(backups_metadata)})
 
-    def _on_restore_action(self, _):
-        # TODO
-        pass
+    def _on_restore_action(self, event: ActionEvent):
+        """Restore a snapshot referenced by its id.
+
+        Steps:
+        - stop client traffic
+        - stop all units
+        - backup local state so that we can rollback if anything goes wrong (manual op)
+        - wipe data folders
+        - get snapshot from object storage, save in data folder
+        - restart units
+        - cleanup leftover files
+        - notify clients
+        """
+        failure_conditions = [
+            (not self.charm.unit.is_leader(), "Action must be ran on the application leader"),
+            (
+                not self.charm.state.cluster.s3_credentials,
+                "Cluster needs an access to an object storage to make a backup",
+            ),
+            (
+                not (id_to_restore := event.params.get("backup-id", "")),
+                "No backup id to restore provided",
+            ),
+            (
+                not self.backup_manager.is_snapshot_in_bucket(id_to_restore),
+                "Backup id not found in storage object",
+            ),
+            (
+                bool(self.charm.state.cluster.id_to_restore),
+                "A snapshot restore is currently ongoing",
+            ),
+        ]
+
+        for check, msg in failure_conditions:
+            if check:
+                logging.error(msg)
+                event.set_results({"error": msg})
+                event.fail(msg)
+                return
+
+        self.charm.state.cluster.update(
+            {
+                "id-to-restore": id_to_restore,
+                "restore-instruction": RestoreStep.NOT_STARTED.value,
+            }
+        )
+        self.charm.disconnect_clients()
+
+        event.log(f"Beginning restore flow for snapshot {id_to_restore}")
+
+    def _restore_event_dispatch(self, event: RelationEvent):
+        """Dispatch restore event to the proper method."""
+        cluster_state = self.charm.state.cluster
+        unit_state = self.charm.state.unit_server
```

Comment:
**question:** Was this just for readability? I'm not suggesting you change it, but it does help IMO having to always remember that 'app state' is the full `self.charm.state.cluster.foo`, rather than having it named differently around the codebase.
This is a VERY minor thing though, just sharing the thought.

Reply:
No no, you are right. This is a leftover from the `restore` peer relation where I needed a clear visual clue about which peer relation I was referring too. Inlined in d6e4c979e378